### ISSUE TITLE
New Rule: ban unsafe declaration merging

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#5275]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

I've seen people doing the interface into class declaration merging pattern a few times (like in #5273) and I don't think people understand how unsafe it can be.

I propose a lint rule which detects instances of unsafe declaration merging and warns against it.

### Fail Cases

```typescript
interface Foo {}
class Foo {}
```

### Pass Cases

```typescript
interface Foo {}
class Bar implements Foo {}

namespace Foo2 {}
namespace Foo2 {}
enum Foo2 {}

namespace Foo3 {}
function Foo3() {}
```
